### PR TITLE
Add Ssc, Txp, and Update to cardano-chain

### PR DESCRIPTION
**Pull request by ruhatch**
I'm putting these three together, because I want things to get through a little faster.

The Ssc modules will disappear completely (and have on my branch), so please completely ignore them. The other two will need cleaning up, but again this is just the first review, so make it light!

### CODE REVIEW

**Review by redxaxder — APPROVED** [review]

File: src/Cardano/Chain/Txp/Tx.hs [hunk 1]
```diff
@@ -0,0 +1,256 @@
+{-# LANGUAGE DeriveGeneric     #-}
+{-# LANGUAGE FlexibleContexts  #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE TemplateHaskell   #-}
+{-# LANGUAGE TypeApplications  #-}
+
+module Cardano.Chain.Txp.Tx
+       ( Tx (..)
```
Comment (mdimjasevic) [inline src/Cardano/Chain/Txp/Tx.hs:8]: Is it OK to export an unsafe data constructor for the `Tx` type? I'd prefer smart constructors.

File: src/Cardano/Chain/Txp/Tx.hs [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+          ]
+        )
+
+instance NFData TxIn
+
+isTxInUnknown :: TxIn -> Bool
+isTxInUnknown (TxInUnknown _ _) = True
+isTxInUnknown _                 = False
+
+txInFromText :: Text -> Either Text TxIn
```
Comment (mdimjasevic) [inline src/Cardano/Chain/Txp/Tx.hs:197]: With a separate (unrelated?) PR you are removing all occurrences of `Either Text`. Why keep it here, i.e., what is different here that justifies keeping it?

File: src/Cardano/Chain/Txp/TxAux.hs [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+txaF :: Format r (TxAux -> r)
+txaF = later $ \(TxAux tx w) ->
+  bprint (build % "\n" % "witnesses: " % listJsonIndent 4) tx w
+
+instance B.Buildable TxAux where
+  build = bprint txaF
+
+-- | Check that a 'TxAux' is internally valid (checks that its 'Tx' is valid via
+--   'checkTx'). Does not check the witness.
+checkTxAux :: MonadError Text m => TxAux -> m ()
```
Comment (mdimjasevic) [inline src/Cardano/Chain/Txp/TxAux.hs:42]: In a PR you are removing all `MonadError Text` occurrences. Why keep it here?

File: src/Cardano/Chain/Txp/TxPayload.hs [hunk 4]
```diff
@@ -0,0 +1,54 @@
+{-# LANGUAGE DeriveGeneric    #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE TemplateHaskell  #-}
+
+module Cardano.Chain.Txp.TxPayload
+       ( TxPayload (..)
```
Comment (mdimjasevic) [inline src/Cardano/Chain/Txp/TxPayload.hs:6]: What if only the type constructor was exported for `TxPayload`? Is its data constructor used anywhere else? I guess `mkTxPayload`, which is exported anyway, could be used instead.

File: src/Cardano/Chain/Update/Payload.hs [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+instance Bi Payload where
+  encode payload = encodeListLen 2 <> encode (payloadProposal payload) <> encode
+    (payloadVotes payload)
+
+  decode = do
+    enforceSize "Update.Payload" 2
+    Payload <$> decode <*> decode
+
+checkPayload :: MonadError Text m => ProtocolMagic -> Payload -> m ()
```
Comment (mdimjasevic) [inline src/Cardano/Chain/Update/Payload.hs:47]: `MonadError Text` here, which is to be avoided.

File: src/Cardano/Chain/Update/SoftwareVersion.hs [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+
+instance Bi SoftwareVersion where
+  encode sv = encodeListLen 2 <> encode (svAppName sv) <> encode (svNumber sv)
+
+  decode = do
+    enforceSize "SoftwareVersion" 2
+    SoftwareVersion <$> decode <*> decode
+
+-- | A software version is valid iff its application name is valid
+checkSoftwareVersion :: MonadError Text m => SoftwareVersion -> m ()
```
Comment (mdimjasevic) [inline src/Cardano/Chain/Update/SoftwareVersion.hs:51]: Another `MonadError Text`...

File: src/Cardano/Chain/Update/SystemTag.hs [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+deriveToJSON defaultOptions ''SystemTag
+
+instance Bi SystemTag where
+  encode = encode . getSystemTag
+  decode = SystemTag <$> decode
+
+systemTagMaxLength :: Integral i => i
+systemTagMaxLength = 10
+
+checkSystemTag :: MonadError Text m => SystemTag -> m ()
```
Comment (mdimjasevic) [inline src/Cardano/Chain/Update/SystemTag.hs:51]: This is another `MonadError Text`.

**Pull request merged**